silo-core: fractions overflow

IhorSF opened the pull request.

0xNeoRacer commented on silo-core/contracts/lib/SiloLendingLib.sol:
```diff
@@ -428,9 +428,19 @@ library SiloLendingLib {
 
-        $.totalAssets[ISilo.AssetType.Collateral] =
-            $.totalAssets[ISilo.AssetType.Collateral] + integralInterest - integralRevenue;
+        (,uint256 newTotalDebtAssets) = $.totalAssets[ISilo.AssetType.Debt].tryAdd(integralInterest);
```
when calculating fraction, we have overflow check on `total * percent` and then we return 0. so:
- if we overflow we do `+= 0`, 
- if we not overflow we doing this: `+= (total * percent) % 1e18`, so is it possible to overflow on `$.totalAssets[ISilo.AssetType.Debt] += integralInterest`?

IhorSF replied:
Outdated

0xNeoRacer commented on silo-core/contracts/lib/SiloLendingLib.sol:
```diff
@@ -428,9 +428,19 @@ library SiloLendingLib {
+            $.totalAssets[ISilo.AssetType.Debt] = newTotalDebtAssets;
+        }
+
+        (, uint256 accruedCollateral) = integralInterest.trySub(integralRevenue);
```
this looks not right to me, 
- we know it is possible to have `+ 0 - 1` so we will have to sub `1`
- this values are % of debt, and debt possible only if we have collateral and `collateral >= debt`, so in extream case, when all fees are 100%, we could have `collateral + integralInterest - integralRevenue` => `1 + 0 - 1 = 0`. 
- by doing `trySub` we basically ignoring this `-1` so we increasing collateral and this sound odd to me.

shadowyedd replied:
outdated

0xNeoRacer commented:
is this enough of the proof that we will not overflow https://github.com/silo-finance/silo-contracts-v2/pull/1174/files#diff-0866cd6dd702e5522152e1f2bf68fbfc35e4a4945698e362bbc58519f7b994b2R29 ?

IhorSF commented:
> is this enough of the proof that we will not overflow https://github.com/silo-finance/silo-contracts-v2/pull/1174/files#diff-0866cd6dd702e5522152e1f2bf68fbfc35e4a4945698e362bbc58519f7b994b2R29 ?

Added a test that show that we can underflow when calculating a cap [see the commit](https://github.com/silo-finance/silo-contracts-v2/pull/1174/commits/6147a2ef3469a8b4fa65333bd9225e8bb3698a1d). The cap == type(uint256).max when it should be 0.

Added a test that shows that applyFractions fn can revert because of the overflow [this commit](https://github.com/silo-finance/silo-contracts-v2/pull/1174/commits/2091700bf96af6d9ffbe3cd9171ca38296ccefe3)

Idk if the scenarios are realistic for the normally operated silos (probably no), but technically, issues are present.

shadowyedd commented on silo-core/contracts/lib/SiloLendingLib.sol:
```diff
@@ -428,9 +428,19 @@ library SiloLendingLib {
         totalFees = _totalFees + integralRevenue;
```
if ($.totalAssets[ISilo.AssetType.Collateral] == type(uint256).max) return (_accruedInterest, _totalFees);

shadowyedd replied:
$.totalAssets[ISilo.AssetType.Collateral] == 0

IhorSF replied:
Done

shadowyedd commented on silo-core/contracts/lib/SiloLendingLib.sol:
```diff
@@ -428,9 +428,19 @@ library SiloLendingLib {
 
-        $.totalAssets[ISilo.AssetType.Collateral] =
-            $.totalAssets[ISilo.AssetType.Collateral] + integralInterest - integralRevenue;
+        (,uint256 newTotalDebtAssets) = $.totalAssets[ISilo.AssetType.Debt].tryAdd(integralInterest);
```
outdated

shadowyedd commented on silo-core/contracts/lib/SiloLendingLib.sol:
```diff
@@ -430,7 +436,7 @@ library SiloLendingLib {
 
-        $.totalAssets[ISilo.AssetType.Collateral] =
-            $.totalAssets[ISilo.AssetType.Collateral] + integralInterest - integralRevenue;
+        (, uint256 accruedCollateral) = integralInterest.trySub(integralRevenue);
```
We can keep original code here as well because totalCollateralAssets > 0

0xNeoRacer commented on silo-core/contracts/lib/SiloLendingLib.sol:
```diff
@@ -406,10 +406,16 @@ library SiloLendingLib {
+        uint256 totalCollateralAssets = $.totalAssets[ISilo.AssetType.Collateral];
+
+        if (totalCollateralAssets == type(uint256).max || totalCollateralAssets == 0) {
+            return (_accruedInterest, _totalFees);
```
```suggestion
            return (0, 0);
```

this is more clear for me and future proof.

0xNeoRacer replied:
ah right, then I would move this check to "core", so to the place where we accrue interest and I would return 0, then all other methods that uses interest will have 0.

shadowyedd replied:
I'd keep named return values even if only for code readability

0xNeoRacer replied:
> I'd keep named return values even if only for code readability
it is not about names, my suggestion is to return 0, if we stay with variables we canreturn non zeros (maybe they are already > 0). so I would put this edge cases (this `if`s) in method where we accrue interest, not where we apply fracions. If we do that in accrue interest, then all other methods that uses interesy will get 0 and all places will be in sync.

IhorSF replied:
An interest is calculated based on the debt assets. If we overflow, `getDebtAmountsWithInterest` will return 0. I think it is enough.

shadowyedd replied:
I vote not to do it.

0xNeoRacer replied:
my intention was NOT to modify `applyFraction` but add this edge cases to accrue interest method and return 0 there.

shadowyedd replied:
// accrueInterestForAsset should never revert so we check edge cases for revert and do early return instead of checking each calculation individually

0xNeoRacer commented:
approved but I assume this two comments will be applied:
- https://github.com/silo-finance/silo-contracts-v2/pull/1159#discussion_r2041826018
- https://github.com/silo-finance/silo-contracts-v2/pull/1159#discussion_r2041821162

shadowyedd commented on silo-core/contracts/lib/SiloLendingLib.sol:
```diff
@@ -406,10 +406,18 @@ library SiloLendingLib {
+        uint256 totalCollateralAssets = $.totalAssets[ISilo.AssetType.Collateral];
+
+        // `accrueInterestForAsset` should never revert,
+        // so we check edge cases for revert and do early return instead of checking each calculation individually
```
```suggestion
        // so we check edge cases for revert and do early return instead of checking each calculation individually for underflow/overflow
```
